crates/uv-build-backend/src/lib.rs

```diff
@@ -336,6 +344,51 @@ pub fn metadata(source_tree: &Path, metadata_directory: &Path) -> Result<String,
     Ok(dist_info_dir)
 }
 
+/// PEP 517 requires that the metadata directory from the prepare metadata call is identical to the
+/// build wheel call. This method performs a prudence check that `METADATA` and `entry_points.txt`
+/// match.
+fn check_metadata_directory(
+    source_tree: &Path,
+    metadata_directory: Option<&Path>,
+    pyproject_toml: &PyProjectToml,
+) -> Result<(), Error> {
+    let Some(metadata_directory) = metadata_directory else {
+        return Ok(());
+    };
+
+    let dist_info_dir = format!(
+        "{}-{}.dist-info",
+        pyproject_toml.name().as_dist_info_name(),
+        pyproject_toml.version()
+    );
+
+    let metadata = pyproject_toml
+        .to_metadata(source_tree)?
+        .core_metadata_format();
+    if fs_err::read_to_string(metadata_directory.join(&dist_info_dir).join("METADATA"))? != metadata
+    {
+        return Err(Error::InconsistentSteps("METADATA"));
```

Comment:
If we expect this to actually happen, it would be more useful for the error to indicate where the mismatch is. But I'm fine with merging as-is and only adding that if it proves useful. (Since it's probably somewhat annoying to implement.)

Reply:
I see these more as `debug_assert!` (i don't see a way how we could violate this invariant), so I've kept the reporting minimal.

Reply:
Yeah that makes sense!